jobsdb/jobsdb.go

```diff
@@ -124,51 +124,84 @@ type statTags struct {
 
 var getTimeNowFunc = time.Now
 
+// Tx is a wrapper around sql.Tx that supports registering and executing
+// post-commit actions, a.k.a. success listeners.
+type Tx struct {
+	*sql.Tx
+	successListeners []func()
+}
+
+// AddSuccessListener registers a listener to be executed after the transaction has been committed successfully.
+func (tx *Tx) AddSuccessListener(listener func()) {
+	tx.successListeners = append(tx.successListeners, listener)
+}
+
+// Commit commits the transaction and executes all listeners.
+func (tx *Tx) Commit() error {
+	err := tx.Tx.Commit()
+	if err == nil {
```

Comment:
Any reason why we are not checking the error early and returning?

Reply:
In the future we might want to introduce other listeners as well, such as `errorListeners` and `completionListeners`. In such a case returning early wouldn't be relevant.
Now it can be done, but I preferred having one less loc instead :)

```
if err != nil {
	return err
}
for _, successListener := range tx.successListeners {
	successListener()
}
return nil
```